prover/witness_generator/src/recursion_tip.rs

```diff
@@ -0,0 +1,341 @@
+            .observe(blob_save_started_at.elapsed());
+
+        let mut prover_connection = self.prover_connection_pool.connection().await.unwrap();
+        let mut transaction = prover_connection.start_transaction().await.unwrap();
+        let protocol_version_id = transaction
+            .fri_witness_generator_dal()
+            .protocol_version_for_l1_batch(job_id)
+            .await;
+        transaction
+            .fri_prover_jobs_dal()
+            .insert_prover_job(
+                job_id,
+                255,
+                0,
+                0,
+                AggregationRound::RecursionTip,
+                &recursion_tip_circuit_blob_url,
+                false,
+                protocol_version_id,
+            )
+            .await;
+
+        transaction
+            .fri_witness_generator_dal()
+            .mark_recursion_tip_job_as_successful(job_id, started_at.elapsed())
+            .await;
+
+        transaction.commit().await.unwrap();
+        Ok(())
+    }
+
+    fn max_attempts(&self) -> u32 {
+        self.config.max_attempts
+    }
+
+    async fn get_job_attempts(&self, job_id: &L1BatchNumber) -> anyhow::Result<u32> {
+        let mut prover_storage = self
+            .prover_connection_pool
+            .connection()
+            .await
+            .context("failed to acquire DB connection for RecursionTipWitnessGenerator")?;
+        prover_storage
+            .fri_witness_generator_dal()
+            .get_recursion_tip_witness_job_attempts(*job_id)
+            .await
+            .map(|attempts| attempts.unwrap_or(0))
+            .context("failed to get job attempts for RecursionTipWitnessGenerator")
+    }
+}
+
+pub async fn prepare_job(
+    l1_batch_number: L1BatchNumber,
+    final_node_proof_job_ids: Vec<(u8, u32)>,
+    object_store: &dyn ObjectStore,
+) -> anyhow::Result<RecursionTipWitnessGeneratorJob> {
+    let started_at = Instant::now();
+    let recursion_tip_proofs =
+        load_proofs_for_recursion_tip(final_node_proof_job_ids, object_store).await?;
+    WITNESS_GENERATOR_METRICS.blob_fetch_time[&AggregationRound::RecursionTip.into()]
+        .observe(started_at.elapsed());
+
+    let keystore = Keystore::default();
+    let node_vk = keystore
+        .load_recursive_layer_verification_key(
+            ZkSyncRecursionLayerStorageType::NodeLayerCircuit as u8,
+        )
+        .context("get_recursive_layer_vk_for_circuit_type()")?;
+
+    let node_layer_vk_commitment = compute_node_vk_commitment(node_vk.clone());
+
+    let mut recursion_queues = vec![];
+    for circuit_id in BaseLayerCircuitType::as_iter_u8() {
+        let key = ClosedFormInputKey {
+            block_number: l1_batch_number,
+            circuit_id,
+        };
+        let ClosedFormInputWrapper(_, recursion_queue) = object_store.get(key).await.unwrap();
+        recursion_queues.push((circuit_id, recursion_queue));
+    }
+
+    assert!(
+        RECURSION_TIP_ARITY >= recursion_queues.len(),
```

Comment:
also maybe add a small comment explaining the concept of the ARITY here (as maximum amount of proofs that this recursion tip can handle)